File: apps/server-admin/src/lib/ddns/providers/godaddy.ts
import type { DDNSProviderContext, DDNSProviderDefinition, DDNSUpdateResult } from "../types";
import {
  ddnsProviderT,
  getTimeoutMs,
  splitDomain,
  toPositiveInt,
  updateDualStack,
} from "./helpers";

const godaddyT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) => ddnsProviderT("godaddy", key, params);
const commonT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) => ddnsProviderT("common", key, params);

export const godaddyProvider: DDNSProviderDefinition = {
  name: "godaddy",
  label: "GoDaddy",
  fields: [
    { key: "api_key", label: "API Key", type: "text", placeholder: "GoDaddy API Key", required: true },
    { key: "api_secret", label: "API Secret", type: "password", placeholder: "GoDaddy API Secret", required: true },
    { key: "root_domain", label: commonT("fields.root_domain.label"), type: "text", placeholder: "example.com", required: true },
    { key: "domain", label: commonT("fields.domain.label"), type: "text", placeholder: "home.example.com", required: true },
    { key: "ttl", label: "TTL", type: "text", placeholder: "600", required: false, description: commonT("fields.ttl.description", { seconds: 600 }) },
  ],
};

export async function godaddyUpdate(
  { config, http }: DDNSProviderContext,
  ipv4: string | null,
  ipv6: string | null,
): Promise<DDNSUpdateResult> {
  const { api_key, api_secret, root_domain, domain } = config;
  if (!api_key || !api_secret || !root_domain || !domain) {
    return { success: false, message: godaddyT("configIncomplete") };
  }

  const ttl = toPositiveInt(config.ttl, 600);
  const parsed = splitDomain(domain, root_domain);

  return updateDualStack("GoDaddy", ipv4, ipv6, async (recordType, ip) => {
    const response = await http.fetch(
      `https://api.godaddy.com/v1/domains/${encodeURIComponent(parsed.rootDomain)}/records/${recordType}/${encodeURIComponent(parsed.recordName)}`,
      {
        method: "PUT",
        headers: {
          Authorization: `sso-key ${api_key}:${api_secret}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify([
          {
            data: ip,
            name: parsed.recordName,
            ttl,
            type: recordType,
          },
        ]),
        signal: AbortSignal.timeout(getTimeoutMs()),
      },
    );

    if (!response.ok) {
      const text = await response.text();
      throw new Error(
        godaddyT("updateFailedWithStatus", {
          status: response.status,
          detail: text || godaddyT("updateFailed"),
        }),
      );
    }
  });
}

File: apps/server-admin-rs/src/ddns/routes/providers/godaddy.rs
use super::*;

pub(in crate::ddns::routes) async fn update_godaddy(
    translator: &Translator,
    config: &HashMap<String, String>,
    http_options: &DDNSHttpClientOptions,
    ipv4: Option<&str>,
    ipv6: Option<&str>,
) -> anyhow::Result<DDNSProviderUpdateResult> {
    let api_key = config_value(config, "api_key");
    let api_secret = config_value(config, "api_secret");
    let root_domain = config_value(config, "root_domain");
    let domain = config_value(config, "domain");
    if api_key.is_empty() || api_secret.is_empty() || root_domain.is_empty() || domain.is_empty() {
        return Ok(provider_failure(ddns_text(
            translator,
            "providers.godaddy.configIncomplete",
            &[],
        )));
    }
    let ttl = positive_i64(config.get("ttl"), 600);
    let parsed = split_domain(translator, &domain, &root_domain)?;
    let client = ddns_http_client(translator, http_options)?;
    let provider_label_text = provider_label(Some("godaddy"), translator);

    update_dual_stack(
        translator,
        &provider_label_text,
        ipv4,
        ipv6,
        |record_type, ip| {
            let client = client.clone();
            let api_key = api_key.clone();
            let api_secret = api_secret.clone();
            let root_domain = parsed.root_domain.clone();
            let record_name = parsed.record_name.clone();
            async move {
                let response = client
                    .put(format!(
                        "https://api.godaddy.com/v1/domains/{}/records/{}/{}",
                        url_encode_component(&root_domain),
                        record_type,
                        url_encode_component(&record_name)
                    ))
                    .header(
                        reqwest::header::AUTHORIZATION,
                        format!("sso-key {api_key}:{api_secret}"),
                    )
                    .json(&json!([{
                        "data": ip,
                        "name": record_name,
                        "ttl": ttl,
                        "type": record_type
                    }]))
                    .send()
                    .await?;
                let status = response.status();
                let text = response_text(response).await?;
                if status.is_success() {
                    Ok(())
                } else {
                    Err(anyhow::anyhow!(ddns_text(
                        translator,
                        "providers.godaddy.updateFailedWithStatus",
                        &[
                            ("status", status.as_u16().to_string()),
                            (
                                "detail",
                                if text.is_empty() {
                                    ddns_text(translator, "providers.godaddy.updateFailed", &[])
                                } else {
                                    text
                                },
                            ),
                        ],
                    )))
                }
            }
        },
    )
    .await
}
